### ISSUE TITLE
2757 - PartialByKeys

### ISSUE THREAD

**Issue by joonseolee** [type-challenges/type-challenges#5395]
```ts
type PartialByKeys<T extends {}, U = keyof T> = 
  Omit<Partial<Pick<T, U & keyof T>> & Omit<T, U & keyof T>, never>;
```

**Comment by k1ngbanana** [comment]
can u tell me what's the Omit<type, never> use for? I think there is not case that never can be a key..but when i remove this. the answer goes wrong.

**Comment by andrewrosss** [comment]
For anyone finding this solution wondering what the purpose of the `Omit<T, never>` is doing: it's basically just creating a new type which is a single object with all the properties/values from the two intersected types.

In particular, using`Omit<T, never>` is just a slick way of implementing the `Copy` or `Merge` types that are common in most of the other solutions.

Omit is defined as:

```ts
type Omit<T, K extends keyof any> = Pick<T, Exclude<keyof T, K>>;
```

When `K` (in the definition of `Omit`, above) is `never`, it's the same thing as doing:

```ts
Pick<T, keyof T>
```

and `Pick`'s implementation looks like:

```ts
type Pick<T, K> = { [P in K]: T[P]; }
```

So, `Pick<T, keyof T>` is the same as doing `{ [P in keyof T]: T[P] }`, which just produces a new type where all the properties/values have been aggregated into a single object.

To sum up, these are all equivalent:

- This solution (using `Omit`):
  ```ts
  type PartialByKeys<T extends {}, U = keyof T> = 
    Omit<Partial<Pick<T, U & keyof T>> & Omit<T, U & keyof T>, never>;
  ```
- Using `Pick`:
  ```ts
  type PartialByKeys<T extends {}, U = keyof T> = 
    Pick<Partial<Pick<T, U & keyof T>> & Omit<T, U & keyof T>, keyof T>;
  ```

- A "custom" copy/merge helper type:
  ```ts
  type _Copy<T> = { [K in keyof T]: T[K] };
  type PartialByKeys<T extends {}, U = keyof T> = 
    _Copy<Partial<Pick<T, U & keyof T>> & Omit<T, U & keyof T>>;
  ```

**Comment by adoin** [comment]
Can any one tell me Why was my solution `type PartialByKeys<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>
` not working?

**Comment by HubooDeclan** [comment]
> Can any one tell me Why was my solution `type PartialByKeys<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>> ` not working?

Just because if you hover over the result of using your solution, the utility types are not expanded so what you need to do is create a copy of the resulting type. Explained in the message above yours how this is done.

<img width="454" alt="Screenshot 2023-05-15 at 09 10 28" src="https://github.com/type-challenges/type-challenges/assets/110537307/a7d72ccd-6289-4e12-a727-d4be16530a81">

**Comment by manyuemeiquqi** [comment]
@HubooDeclan  
It's cool
Can you tell me how to expand type infomation in ide
I hover type only expand the following
![image](https://github.com/type-challenges/type-challenges/assets/54520846/923aad77-36ac-4ce0-8807-2d3722ae730d)

**Comment by HubooDeclan** [comment]
@manyuemeiquqi 

You need to expand the result, you can do this by making a new copy of the type with the same keys and types. In the above example, I've used the following.

`Omit<RESULT, never>`

Or you can use these helper functions

```
/**
 * Expand a type recusively. Makes types much nicer on hover ooooft
 */
export type ExpandRecursively<T> = T extends object
  ? T extends infer O
    ? { [K in keyof O]: ExpandRecursively<O[K]> }
    : never
  : T

/** Expand a type */
export type Expand<T> = { [K in keyof T]: T[K] }

/** Expand a type */
export type Expand<T> = Omit<T, never>
```

**Comment by iwebexpert** [comment]
Good solution! I was able to simplify it a little. 
Maybe this will be useful:
```ts
type PartialByKeys<T, K extends keyof T = keyof T> = Omit<Partial<Pick<T, K>> & Omit<T, K> , never>
```
